jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,369 @@
+            throw new IllegalArgumentException("invalid maxSuspend duration");
+        this.maxSuspend = maxSuspend;
+    }
+
+    /**
+     * @return the max priority of suspended requests
+     */
+    @ManagedAttribute("The maximum priority of suspended requests")
+    public int getMaxPriority()
+    {
+        return maxPriority;
+    }
+
+    /**
+     * <p>Sets the max priority that suspended requests may have.</p>
+     * <p>The value must be {@code 0} or greater, and typically only
+     * few priorities are necessary, say from {@code 0} to {@code 4}
+     * for 5 different priorities.</p>
+     *
+     * @param maxPriority the max priority of suspended requests
+     * @see #getPriority(Request)
+     */
+    public void setMaxPriority(int maxPriority)
+    {
+        if (maxPriority < 0)
+            throw new IllegalArgumentException("invalid maxPriority");
+        this.maxPriority = maxPriority;
+    }
+
+    @ManagedAttribute("The number of suspended requests")
+    public long getSuspendedRequests()
+    {
+        return queues.stream()
+            .mapToLong(Queue::size)
+            .sum();
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        timeouts = new Timeouts(getServer().getScheduler());
+        addBean(timeouts);
+
+        int maxRequests = getMaxRequests();
+        if (maxRequests <= 0)
+        {
+            ThreadPool threadPool = getServer().getThreadPool();
+            if (threadPool instanceof ThreadPool.SizedThreadPool sized)
+                maxRequests = sized.getMaxThreads() / 2;
+            else
+                maxRequests = ProcessorUtils.availableProcessors();
+            setMaxRequests(maxRequests);
+        }
+        semaphore = new Semaphore(maxRequests);
+
+        int maxPriority = getMaxPriority();
+        int capacity = maxPriority + 1;
+        queues = new ArrayList<>(capacity);
+        for (int i = 0; i < capacity; ++i)
+        {
+            queues.add(new ConcurrentLinkedQueue<>());
+        }
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} initialized maxRequests={} maxPriority={}", this, maxRequests, maxPriority);
+
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        removeBean(timeouts);
+        timeouts.destroy();
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} handling {}", this, request);
+        return process(request, response, callback);
+    }
+
+    private boolean process(Request request, Response response, Callback callback) throws Exception
+    {
+        if (semaphore.tryAcquire())
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} forwarding {}", this, request);
+            request.addHttpStreamWrapper(stream -> new Resumer(stream, request));
```

Comment:
At the very least we need to document that "concurrent request" means not just during handling, but all the way through to the completion of the callback.

However, I'm not convinced this will always be the best solution if there are asynchronous handlers involved (and any webapp that serves static content might have an async handler).

Would it be too much to have the option to resume either on completion or on return from handling?

Reply:
I'd stay minimal for this iteration.

Async handlers may do all kind of things, for example consume CPU in a tight loop while processing large amounts of data, so I am not sure I want to allow other requests in, even if the handler returned `true`.

I chose the most conservative approach for now, we can relax later if there is request for it.

Reply:
Ok, but document please